### PR TITLE
Large grammar test. Consumes all memory in the known universe

### PR DESCRIPTION
**Pull request by IzakMarais**
Enjoy.

### CODE REVIEW

File: relapse/mem/testproto/Makefile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+gengo:
```
Comment (awalterschulze) [inline relapse/mem/testproto/Makefile]: This file is not anonymized yet.

File: relapse/mem/testproto/Banana.proto [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    option (gogoproto.description) = true;
+    optional int64 BananaTime = 1 [(gogoproto.nullable) = false];
+    optional string Pulpy = 2 [(gogoproto.nullable) = false];
+    optional bool IsPulpySet = 3 [(gogoproto.nullable) = false];
+    optional string Yummy = 4 [(gogoproto.nullable) = false];
+    optional bool IsYummySet = 5 [(gogoproto.nullable) = false];
+    optional string Chewy = 6 [(gogoproto.nullable) = false];
+    optional bool IsChewySet = 7 [(gogoproto.nullable) = false];
+    optional string Squishy = 8 [(gogoproto.nullable) = false];
+    optional bool IsSquishySet = 9 [(gogoproto.nullable) = false];
```
Comment (awalterschulze) [inline relapse/mem/testproto/Banana.proto]: I love the BananaTuple :)

Comment (IzakMarais) [inline relapse/mem/testproto/Banana.proto]: This is a nice side effect:

```golang
type randyBanana interface {
	Float32() float32
	Float64() float64
	Int63() int64
	Int31() int32
	Uint32() uint32
	Intn(n int) int
}
```

Comment (awalterschulze) [inline relapse/mem/testproto/Banana.proto]: Ha ha :)

**Comment by awalterschulze** [comment]
I just fixed the master branch by updating gogoprotobuf and protoc to version 3.3

**Comment by awalterschulze** [comment]
Ok perfect.  I am taking a look.  I don't know how much I'll be able to do today, but this is something I can work with.

I have some ideas on fixing this with simplification rules in the relapse/funcs/simplify.go , but I don't know if that will be enough.

When its fixed we'll have to think about how to clean this up for merging, but this is super helpful.  Thanks for all your efforts.

**Comment by IzakMarais** [comment]
Excellent, thanks!

**Comment by awalterschulze** [comment]
Is this query really supposed to match the input tuple?

**Comment by awalterschulze** [comment]
This test passes now with the following modification

```go
    tup.Squishy = "422"
    tup.Brown = "02"
    tup.Peel = uint32(12871)
    tup.Rotten = uint32(40111)
```

**Comment by awalterschulze** [comment]
```diff
diff --git a/relapse/funcs/funcs-gen/main.go b/relapse/funcs/funcs-gen/main.go
index 20fafe2..1857556 100644
--- a/relapse/funcs/funcs-gen/main.go
+++ b/relapse/funcs/funcs-gen/main.go
@@ -16,8 +16,9 @@
 package main
 
 import (
-	"github.com/katydid/katydid/gen"
 	"strings"
+
+	"github.com/katydid/katydid/gen"
 )
 
 const compareStr = `
@@ -142,13 +143,14 @@ func (this *listOf{{.FuncType}}) String() string {
 	for i := range this.List {
 		ss[i] = sprint(this.List[i])
 	}
-	return "[]{{.GoType}}{" + strings.Join(ss, ",") + "}"
+	return "[]{{.Type}}{" + strings.Join(ss, ",") + "}"
 }
 
 func (this *listOf{{.FuncType}}) IsListOf() {}
 `
 
 type list struct {
+	Type     string
 	CType    string
 	FuncType string
 	GoType   string
@@ -391,6 +393,9 @@ type inSet{{.Name}} struct {
 }
 
 func (this *inSet{{.Name}}) Init() error {
+	if this.set != nil {
+		return nil
+	}
 	l, err := this.List.Eval()
 	if err != nil {
 		return err
@@ -415,7 +420,7 @@ func init() {
 	Register("contains", new(inSet{{.Name}}))
 }
 
-//Contains{{.Name}} returns a function that checks when the element if contained in the list.
+//Contains{{.Name}} returns a function that checks whether the element is contained in the list.
 func Contains{{.Name}}(element {{.Name}}, list {{.ConstListType}}) Bool {
 	return &inSet{{.Name}}{element, list, nil}
 }
@@ -488,12 +493,12 @@ func main() {
 		&conster{"ListOfBytes", "[][]byte", "%#v", "[]byte"},
 	}, `"fmt"`, `"strings"`, `"reflect"`)
 	gen(listStr, "list.gen.go", []interface{}{
-		&list{"Doubles", "Double", "float64"},
-		&list{"Ints", "Int", "int64"},
-		&list{"Uints", "Uint", "uint64"},
-		&list{"Bools", "Bool", "bool"},
-		&list{"Strings", "String", "string"},
-		&list{"ListOfBytes", "Bytes", "[]byte"},
+		&list{"double", "Doubles", "Double", "float64"},
+		&list{"int", "Ints", "Int", "int64"},
+		&list{"uint", "Uints", "Uint", "uint64"},
+		&list{"bool", "Bools", "Bool", "bool"},
+		&list{"string", "Strings", "String", "string"},
+		&list{"[]byte", "ListOfBytes", "Bytes", "[]byte"},
 	}, `"strings"`)
 	gen(printStr, "print.gen.go", []interface{}{
 		&printer{"Double", "float64"},
diff --git a/relapse/funcs/inset.gen.go b/relapse/funcs/inset.gen.go
index 5741a6b..03e5ec2 100644
--- a/relapse/funcs/inset.gen.go
+++ b/relapse/funcs/inset.gen.go
@@ -8,6 +8,9 @@ type inSetInt struct {
 }
 
 func (this *inSetInt) Init() error {
+	if this.set != nil {
+		return nil
+	}
 	l, err := this.List.Eval()
 	if err != nil {
 		return err
@@ -32,7 +35,7 @@ func init() {
 	Register("contains", new(inSetInt))
 }
 
-//ContainsInt returns a function that checks when the element if contained in the list.
+//ContainsInt returns a function that checks whether the element is contained in the list.
 func ContainsInt(element Int, list ConstInts) Bool {
 	return &inSetInt{element, list, nil}
 }
@@ -44,6 +47,9 @@ type inSetUint struct {
 }
 
 func (this *inSetUint) Init() error {
+	if this.set != nil {
+		return nil
+	}
 	l, err := this.List.Eval()
 	if err != nil {
 		return err
@@ -68,7 +74,7 @@ func init() {
 	Register("contains", new(inSetUint))
 }
 
-//ContainsUint returns a function that checks when the element if contained in the list.
+//ContainsUint returns a function that checks whether the element is contained in the list.
 func ContainsUint(element Uint, list ConstUints) Bool {
 	return &inSetUint{element, list, nil}
 }
@@ -80,6 +86,9 @@ type inSetString struct {
 }
 
 func (this *inSetString) Init() error {
+	if this.set != nil {
+		return nil
+	}
 	l, err := this.List.Eval()
 	if err != nil {
 		return err
@@ -104,7 +113,7 @@ func init() {
 	Register("contains", new(inSetString))
 }
 
-//ContainsString returns a function that checks when the element if contained in the list.
+//ContainsString returns a function that checks whether the element is contained in the list.
 func ContainsString(element String, list ConstStrings) Bool {
 	return &inSetString{element, list, nil}
 }
diff --git a/relapse/funcs/list.gen.go b/relapse/funcs/list.gen.go
index 17e3ae6..6766faa 100644
--- a/relapse/funcs/list.gen.go
+++ b/relapse/funcs/list.gen.go
@@ -31,7 +31,7 @@ func (this *listOfDouble) String() string {
 	for i := range this.List {
 		ss[i] = sprint(this.List[i])
 	}
-	return "[]float64{" + strings.Join(ss, ",") + "}"
+	return "[]double{" + strings.Join(ss, ",") + "}"
 }
 
 func (this *listOfDouble) IsListOf() {}
@@ -62,7 +62,7 @@ func (this *listOfInt) String() string {
 	for i := range this.List {
 		ss[i] = sprint(this.List[i])
 	}
-	return "[]int64{" + strings.Join(ss, ",") + "}"
+	return "[]int{" + strings.Join(ss, ",") + "}"
 }
 
 func (this *listOfInt) IsListOf() {}
@@ -93,7 +93,7 @@ func (this *listOfUint) String() string {
 	for i := range this.List {
 		ss[i] = sprint(this.List[i])
 	}
-	return "[]uint64{" + strings.Join(ss, ",") + "}"
+	return "[]uint{" + strings.Join(ss, ",") + "}"
 }
 
 func (this *listOfUint) IsListOf() {}
diff --git a/relapse/funcs/simplify.go b/relapse/funcs/simplify.go
index f86500c..33f1cfb 100644
--- a/relapse/funcs/simplify.go
+++ b/relapse/funcs/simplify.go
@@ -65,7 +65,8 @@ func Equal(l, r interface{}) bool {
 
 //Simplify simplifies a function logically by eliminating impossible ands, ors with constant true values, double nots, etc.
 func Simplify(f Bool) Bool {
-	if ff, ok := f.(*and); ok {
+	switch ff := f.(type) {
+	case *and:
 		v1 := Simplify(ff.V1)
 		v2 := Simplify(ff.V2)
 		if l, ok := v1.(*constBool); ok {
@@ -195,10 +196,29 @@ func Simplify(f Bool) Bool {
 					}
 				}
 			}
+		case *inSetUint:
+			if vv2, ok := v2.(*inSetUint); ok {
+				if err := vv1.Init(); err != nil {
+					panic(err)
+					return And(v1, v2)
+				}
+				if err := vv2.Init(); err != nil {
+					panic(err)
+					return And(v1, v2)
+				}
+				intersect := make(map[uint64]struct{})
+				for k := range vv1.set {
+					if _, ok := vv2.set[k]; ok {
+						intersect[k] = struct{}{}
+					}
+				}
+				if len(intersect) == 0 {
+					return BoolConst(false)
+				}
+			}
 		}
 		return And(v1, v2)
-	}
-	if ff, ok := f.(*or); ok {
+	case *or:
 		v1 := Simplify(ff.V1)
 		v2 := Simplify(ff.V2)
 		if l, ok := v1.(*constBool); ok {
@@ -229,8 +249,7 @@ func Simplify(f Bool) Bool {
 			}
 		}
 		return Or(v1, v2)
-	}
-	if ff, ok := f.(*not); ok {
+	case *not:
 		v1 := Simplify(ff.V1)
 		if vv, ok := v1.(*not); ok {
 			return vv.V1
@@ -271,6 +290,20 @@ func Simplify(f Bool) Bool {
 			return UintEq(vv.V1, vv.V2)
 		}
 		return Not(v1)
+	case *inSetUint:
+		if err := ff.Init(); err != nil {
+			return f
+		}
+		if len(ff.set) == 0 {
+			return BoolConst(false)
+		}
+		if len(ff.set) == 1 {
+			var one uint64
+			for e := range ff.set {
+				one = e
+			}
+			return UintEq(ff.Elem, UintConst(one))
+		}
 	}
 	return f
 }
diff --git a/relapse/interp/simplify.go b/relapse/interp/simplify.go
index 7eb1771..c5638a9 100644
--- a/relapse/interp/simplify.go
+++ b/relapse/interp/simplify.go
@@ -16,10 +16,12 @@ package interp
 
 import (
 	"fmt"
+
 	"github.com/katydid/katydid/relapse/ast"
 	"github.com/katydid/katydid/relapse/compose"
 	"github.com/katydid/katydid/relapse/funcs"
 	nameexpr "github.com/katydid/katydid/relapse/name"
+	"github.com/katydid/katydid/relapse/parser"
 )
 
 //Simplifier simplifies the patterns of a given grammar.
@@ -122,10 +124,16 @@ func (this *simplifier) simplify(p *ast.Pattern, top bool) *ast.Pattern {
 			//Don't simplify if there is an error to keep this function signature simple.
 			return p
 		}
+		b = funcs.Simplify(b)
 		if funcs.IsFalse(b) {
 			return emptyset
 		}
-		return p
+		e, err := parser.ParseExpr(funcs.Sprint(b))
+		if err != nil {
+			//Don't simplify if there is an error to keep this function signature simple.
+			return p
+		}
+		return ast.NewLeafNode(e)
 	case *ast.Concat:
 		return cRef(simplifyConcat(
 			simp(v.GetLeftPattern()),
diff --git a/relapse/mem/mem_test.go b/relapse/mem/mem_test.go
index cf7edf1..d84096c 100644
--- a/relapse/mem/mem_test.go
+++ b/relapse/mem/mem_test.go
@@ -37,7 +37,10 @@ func TestLargeGrammar(t *testing.T) {
 	parser := katyproto.NewProtoNumParser("testproto", "BananaTuple", testproto.BananaDescription())
 	popr := rand.New(rand.NewSource(time.Now().UnixNano()))
 	tup := testproto.NewPopulatedBananaTuple(popr, true)
-	tup.Peel = uint32(551496705)
+	tup.Squishy = "422"
+	tup.Brown = "02"
+	tup.Peel = uint32(12871)
+	tup.Rotten = uint32(40111)
 
 	data, err := gogoproto.Marshal(tup)
 	if err != nil {
diff --git a/relapse/parser/parse.go b/relapse/parser/parse.go
index 0edcd8e..c8999d0 100644
--- a/relapse/parser/parse.go
+++ b/relapse/parser/parse.go
@@ -16,6 +16,7 @@ package parser
 
 import (
 	"fmt"
+
 	"github.com/katydid/katydid/relapse/ast"
 	"github.com/katydid/katydid/relapse/lexer"
 )
@@ -68,3 +69,9 @@ func (this *Parser) ParseExpr(s string) (res *ast.Expr, err error) {
 	}
 	return gram.GetTopPattern().GetLeafNode().GetExpr(), nil
 }
+
+//ParseExpr returns a parsed expression or error, given a string.
+func ParseExpr(s string) (*ast.Expr, error) {
+	p := NewParser()
+	return p.ParseExpr(s)
+}
```

**Comment by awalterschulze** [comment]
Building the if tree using the addIfExpr method in relapse/mem/ifs.go has exponential behaviour in the worst case.
These new simplification rules curb the growth and results in a much more compact tree and less exponential behaviour.
This test now runs in 0.56s.

I still need to do some more cleanup and think about a fallback strategy for the case where no such simplification is possible.
For now I have posted a diff in case you are in a hurry.

**Comment by awalterschulze** [comment]
I have applied most of the changes, except for the actual simplifications in master.

**Comment by awalterschulze** [comment]
I think to properly solve this in all cases, not have exponential behaviour for cases that don't have simplifications, I'll have to make compileIfExprs memoize (lazily compile) instead of precompute everything.  Just like the mem package is doing for the relapse grammar.

I won't be able to do that today, but you have a diff that will work for now, in case you are in a hurry.

**Comment by awalterschulze** [comment]
I have started to work on a more permanent solution on the https://github.com/katydid/katydid/tree/memifs branch.
Currently the tests including yours passes

```
=== RUN   TestLargeGrammar
--- PASS: TestLargeGrammar (0.12s)
PASS
```

I moved all the compiling behavior, including those in the if expressions to the auto package, since this is actually the only place that is suppose to use it.
This results in some duplicate code, but the hope is that it helps me to clean up the mem package.
The auto package is just there as a benchmark of how fast the mem package can possibly become.
The mem package is the only package should be used in production.

The current implementation of the `ifExprs` https://github.com/katydid/katydid/blob/memifs/relapse/mem/ifs.go#L25-L52 still requires some work.  It still needs:
1) some memoizing of results
2) to not evaluate functions whose results are already known.  For example if there are two functions `(== a)` and `(== b)` then we know that when `(==a)` is true `(==b)` cannot possibly be true.

**Comment by awalterschulze** [comment]
Current benchmarks with no optimizations for the mem package

```
BenchmarkSuite/AndNameTelephonePbnum-8         	  100000	     22220 ns/op
BenchmarkSuite/BridgePepperPbnum-8             	   20000	     90275 ns/op
BenchmarkSuite/BridgePepperAndFountainTargetPbnum-8         	   10000	    126143 ns/op
BenchmarkSuite/ContextPersonPbnum-8                         	  100000	     15596 ns/op
BenchmarkSuite/CorrectNotNamePbnum-8                        	  200000	      9418 ns/op
BenchmarkSuite/EmptyOrNilPbnum-8                            	  100000	     15452 ns/op
BenchmarkSuite/IncorrectNotNamePbnum-8                      	  200000	      6130 ns/op
BenchmarkSuite/LenNamePbnum-8                               	  100000	     12169 ns/op
BenchmarkSuite/ListIndexAddressPbnum-8                      	   50000	     28971 ns/op
BenchmarkSuite/NilNamePbnum-8                               	  500000	      3377 ns/op
BenchmarkSuite/OrNameTelephonePbnum-8                       	  100000	     22052 ns/op
BenchmarkSuite/RecursiveSrcTreePbnum-8                      	   30000	     47944 ns/op
BenchmarkSuite/TypewriterPrisonDaisySledPbnum-8             	   20000	     60607 ns/op
BenchmarkSuite/TypewriterPrisonMapSharkPbnum-8              	   20000	     60824 ns/op
BenchmarkSuite/TypewriterPrisonMenuPaperclipPbnum-8         	   30000	     52212 ns/op
BenchmarkSuite/TypewriterPrisonScarBusStopPbnum-8           	   50000	     40062 ns/op
BenchmarkSuite/TypewriterPrisonSmileLetterPbnum-8           	   30000	     43224 ns/op
```

**Comment by awalterschulze** [comment]
Current benchmarks with no optimizations, but with memoization on the ifExprs for the mem package.

```
BenchmarkSuite/AndNameTelephonePbnum-8         	 2000000	       605 ns/op
BenchmarkSuite/BridgePepperPbnum-8             	  300000	      5679 ns/op
BenchmarkSuite/BridgePepperAndFountainTargetPbnum-8         	  200000	      6005 ns/op
BenchmarkSuite/ContextPersonPbnum-8                         	 2000000	       636 ns/op
BenchmarkSuite/CorrectNotNamePbnum-8                        	 5000000	       281 ns/op
BenchmarkSuite/EmptyOrNilPbnum-8                            	 3000000	       505 ns/op
BenchmarkSuite/IncorrectNotNamePbnum-8                      	 5000000	       274 ns/op
BenchmarkSuite/LenNamePbnum-8                               	 3000000	       470 ns/op
BenchmarkSuite/ListIndexAddressPbnum-8                      	 1000000	      1165 ns/op
BenchmarkSuite/NilNamePbnum-8                               	10000000	       186 ns/op
BenchmarkSuite/OrNameTelephonePbnum-8                       	 2000000	       607 ns/op
BenchmarkSuite/RecursiveSrcTreePbnum-8                      	 1000000	      1431 ns/op
BenchmarkSuite/TypewriterPrisonDaisySledPbnum-8             	  500000	      3401 ns/op
BenchmarkSuite/TypewriterPrisonMapSharkPbnum-8              	  500000	      3321 ns/op
BenchmarkSuite/TypewriterPrisonMenuPaperclipPbnum-8         	  500000	      2616 ns/op
BenchmarkSuite/TypewriterPrisonScarBusStopPbnum-8           	 1000000	      2061 ns/op
BenchmarkSuite/TypewriterPrisonSmileLetterPbnum-8           	  500000	      2340 ns/op
```

**Comment by awalterschulze** [comment]
Current master

```
BenchmarkSuite/AndNameTelephonePbnum-8         	 3000000	       549 ns/op
BenchmarkSuite/BridgePepperPbnum-8             	  300000	      5467 ns/op
BenchmarkSuite/BridgePepperAndFountainTargetPbnum-8         	  200000	      6169 ns/op
BenchmarkSuite/ContextPersonPbnum-8                         	 2000000	       710 ns/op
BenchmarkSuite/CorrectNotNamePbnum-8                        	 5000000	       264 ns/op
BenchmarkSuite/EmptyOrNilPbnum-8                            	 3000000	       469 ns/op
BenchmarkSuite/IncorrectNotNamePbnum-8                      	 5000000	       286 ns/op
BenchmarkSuite/LenNamePbnum-8                               	 3000000	       446 ns/op
BenchmarkSuite/ListIndexAddressPbnum-8                      	 1000000	      1087 ns/op
BenchmarkSuite/NilNamePbnum-8                               	10000000	       183 ns/op
BenchmarkSuite/OrNameTelephonePbnum-8                       	 3000000	       547 ns/op
BenchmarkSuite/RecursiveSrcTreePbnum-8                      	 1000000	      1357 ns/op
BenchmarkSuite/TypewriterPrisonDaisySledPbnum-8             	  500000	      3331 ns/op
BenchmarkSuite/TypewriterPrisonMapSharkPbnum-8              	  500000	      3177 ns/op
BenchmarkSuite/TypewriterPrisonMenuPaperclipPbnum-8         	  500000	      2571 ns/op
BenchmarkSuite/TypewriterPrisonScarBusStopPbnum-8           	 1000000	      1991 ns/op
BenchmarkSuite/TypewriterPrisonSmileLetterPbnum-8           	  500000	      2218 ns/op
```

**Comment by awalterschulze** [comment]
Benchmarks with optimizations

```
BenchmarkSuite/AndNameTelephonePbnum-8         	 3000000	       577 ns/op
BenchmarkSuite/BridgePepperPbnum-8             	  300000	      5602 ns/op
BenchmarkSuite/BridgePepperAndFountainTargetPbnum-8         	  200000	      6030 ns/op
BenchmarkSuite/ContextPersonPbnum-8                         	 2000000	       622 ns/op
BenchmarkSuite/CorrectNotNamePbnum-8                        	 5000000	       257 ns/op
BenchmarkSuite/EmptyOrNilPbnum-8                            	 3000000	       481 ns/op
BenchmarkSuite/IncorrectNotNamePbnum-8                      	 5000000	       267 ns/op
BenchmarkSuite/LenNamePbnum-8                               	 3000000	       469 ns/op
BenchmarkSuite/ListIndexAddressPbnum-8                      	 1000000	      1113 ns/op
BenchmarkSuite/NilNamePbnum-8                               	10000000	       190 ns/op
BenchmarkSuite/OrNameTelephonePbnum-8                       	 3000000	       587 ns/op
BenchmarkSuite/RecursiveSrcTreePbnum-8                      	 1000000	      1405 ns/op
BenchmarkSuite/TypewriterPrisonDaisySledPbnum-8             	  500000	      3384 ns/op
BenchmarkSuite/TypewriterPrisonMapSharkPbnum-8              	  500000	      3244 ns/op
BenchmarkSuite/TypewriterPrisonMenuPaperclipPbnum-8         	  500000	      2540 ns/op
BenchmarkSuite/TypewriterPrisonScarBusStopPbnum-8           	 1000000	      2061 ns/op
BenchmarkSuite/TypewriterPrisonSmileLetterPbnum-8           	  500000	      2302 ns/op
```

**Comment by awalterschulze** [comment]
And the test passes
```
=== RUN   TestLargeGrammar
--- PASS: TestLargeGrammar (0.14s)
```

**Comment by awalterschulze** [comment]
I will merge this pull request
https://github.com/katydid/katydid/pull/4

What I would like from you is to make a new pull request on https://github.com/katydid/testsuite
The reason the testsuite is separate is that it is language agnostic.
It allows me to run the tests on both of the current implementations of relapse, this one and the one in haskell.

I would really appreciate it if could go into this folder an add your test there please
https://github.com/katydid/testsuite/tree/master/relapse/gen-relapse-tests

I would really love to use the BananaTuple in future and the test you created here should be used in all future implementations to make sure they don't have an exponential explosion either.

I am going to close this pull request now, because it is fixed.
Thanks again for reporting :)

**Comment by IzakMarais** [comment]
I will get round to this today. Sorry for the delay; got busy with other things.

**Comment by awalterschulze** [comment]
No problem.  Looking forward to it :)

**Pull request closed without merging**